### PR TITLE
Feature/add dspfparser module

### PR DESCRIPTION
**Pull request by andraasi**
## Description

Add dspfparser module

## Checklist:
- [x] There are tests regarding this feature
- [x] The code follows the Kotlin conventions (run `./gradlew ktlintCheck`)
- [x] The code passes all tests (run `./gradlew check`)
- [ ] There is a specific documentation in the `docs` directory

### CODE REVIEW

File: dspfparser/src/test/resources/_d2.dspf
Comment (lanarimarco) [inline dspfparser/src/test/resources/_d2.dspf:1]: capitalize the name

File: dspfparser/src/test/resources/_fake.dspf
Comment (lanarimarco) [inline dspfparser/src/test/resources/_fake.dspf:1]: capitalize the name

File: dspfparser/src/test/resources/b£1601av.dspf
Comment (lanarimarco) [inline dspfparser/src/test/resources/b£1601av.dspf:1]: capitalize the name

File: dspfparser/src/test/resources/b£1601bv.dspf
Comment (lanarimarco) [inline dspfparser/src/test/resources/b£1601bv.dspf:1]: capitalize the name

**Review by lanarimarco — CHANGES_REQUESTED** [review]
There is something to fix.

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    }
+}
+
+private enum class CurrentContext {
+    FILE,
+    RECORD,
+    FIELD,
+}
+
+internal const val SHOULD_ADD_RELATED_AND_FILE: Boolean = false
```
[the remark this answers is not in the document]
Comment (andraasi) [reply dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt]: For this "switch" and the other similar in DSPFLine, I decided to simply remove them. Its purpose were to make some asserts (asserts concerning "will-come" features, that are currenlty out of scope) skip based on its values. I removed those asserts too.

File: dspfparser/src/test/kotlin/com/smeup/dspfparser/linesprocessor/KeywordsTest.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        assertEquals("SFL", sfl.name)
+    }
+
+    @Test
+    fun constant() {
+        val constant = DSPFKeyword.fromString("'* Constant message! *'")
+        assertEquals("* Constant message! *", constant.name)
+    }
+
+    @Test @Ignore
```
[the remark this answers is not in the document]
Comment (andraasi) [reply dspfparser/src/test/kotlin/com/smeup/dspfparser/linesprocessor/KeywordsTest.kt]: My fault, this was a leftover test, I removed it.

File: dspfparser/src/test/kotlin/com/smeup/dspfparser/linesclassifier/B1601av.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.smeup.dspfparser.linesclassifier
+
+import kotlin.test.Test
+
+internal class B1601av : DSPFSpecificationsLoader("./src/test/resources/b£1601av.dspf") {
+    @Test
+    fun tt() {
```
[the remark this answers is not in the document]
Comment (andraasi) [reply dspfparser/src/test/kotlin/com/smeup/dspfparser/linesclassifier/B1601av.kt]: Forgot to add test asserts. Now added a bare minimum assertions.

File: dspfparser/build.gradle [hunk 4]
```diff
@@ -1,40 +1,99 @@
...
+    }
+}
+
+// deploy
+publishing {
+    publications {
+        mavenJava(MavenPublication) {
+            customizePom(pom)
+            groupId jarikoGroupId
+            artifactId 'dspfarser'
```
Comment (lanarimarco) [inline dspfparser/build.gradle:82]: fix typo

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    companion object {
+        fun fromLines(
+            lines: MutableList<DSPFLine>,
+            name: String? = null
+        ): DSPFSpecifications {
+            return DSPFSpecificationsFactory(name, lines).create()
+        }
+    }
+
+    override fun getRecord(name: String): DSPFRecordSpecifications {
```
[the remark this answers is not in the document]
Comment (lanarimarco) [reply dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt:23]: Is there a test regarding this feature? I think this implementation is wrong

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        return this.records.first { it.declaration.fieldName == name }
+    }
+
+    override fun getFieldsFromRecord(name: String): MutableList<DSPFFieldSpecifications> {
```
[the remark this answers is not in the document]
Comment (lanarimarco) [reply dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt:27]: Is there a test regarding this feature? I think this implementation is wrong

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt [hunk 7]
```diff
@@ -14,24 +14,24 @@ internal fun merge(linesSubstrings: MutableList<DSPFLineSubstrings>): DSPFLineSu
     linesSubstrings.forEach {
         if (it.continuationChar != null) {
             if (continuesAtAnyColumn == null) {
-                count = it.count!!
-                line += it.line!!.removeLast(it.continuationChar!!.toString())
+                count = it.count
+                line += it.line.removeLast(it.continuationChar.toString())
```
Comment (lanarimarco) [inline dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt:18]: Concat using `line +=` could be inefficient. Replace it with `StringBuilder.append` or equivalent.

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt [hunk 8]
```diff
@@ -14,24 +14,24 @@ internal fun merge(linesSubstrings: MutableList<DSPFLineSubstrings>): DSPFLineSu
...
                 conditions += it.condition
             } else {
                 if (continuesAtAnyColumn == true) {
-                    line += it.line!!.trim().removeLast(it.continuationChar!!.toString())
+                    line += it.line.trim().removeLast(it.continuationChar.toString())
```
Comment (lanarimarco) [inline dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt:22]: see above

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt [hunk 9]
```diff
@@ -14,24 +14,24 @@ internal fun merge(linesSubstrings: MutableList<DSPFLineSubstrings>): DSPFLineSu
...
                 }
                 if (continuesAtAnyColumn == false) {
-                    line += it.keywords!!.trim().removeLast(it.continuationChar!!.toString())
+                    line += it.keywords.trim().removeLast(it.continuationChar.toString())
```
Comment (lanarimarco) [inline dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt:25]: see above

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt [hunk 10]
```diff
@@ -14,24 +14,24 @@ internal fun merge(linesSubstrings: MutableList<DSPFLineSubstrings>): DSPFLineSu
...
                 }
             }
             if (it.continuesAtAnyColumn()) continuesAtAnyColumn = true
             if (it.continuesAtColumn45()) continuesAtAnyColumn = false
         } else {
-            if (continuesAtAnyColumn == true) line += it.line!!.trim()
+            if (continuesAtAnyColumn == true) line += it.line.trim()
             if (continuesAtAnyColumn == false) {
-                line += it.keywords!!.trim()
+                line += it.keywords.trim()
```
Comment (lanarimarco) [inline dspfparser/src/main/kotlin/com/smeup/dspfparser/linesprocessor/LinesSubstringsMerger.kt:34]: see above

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFFieldSpecifications.kt [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    override val value: DSPFFieldValue<String>,
+    override val isNumeric: Boolean,
+    override val length: Int? = null,
+    override val precision: Int? = null,
+    override val type: DSPFFieldType,
+    override val x: Int,
+    override val y: Int,
+    override var hasError: Boolean = false
+) : DSPFField {
+    val related: MutableList<DSPFLine> = mutableListOf()
```
Comment (lanarimarco) [inline dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFFieldSpecifications.kt:21]: This property is no longer used, why?

File: dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+import com.smeup.dspfparser.domain.DSPF
+import com.smeup.dspfparser.linesprocessor.DSPFLine
+import com.smeup.rpgparser.interpreter.DbField
+import kotlinx.serialization.Serializable
+
+@Serializable
+internal data class DSPFSpecifications(
+    override val name: String?
+) : DSPF {
+    val file: DSPFFileSpecifications = DSPFFileSpecifications.create()
```
Comment (lanarimarco) [inline dspfparser/src/main/kotlin/com/smeup/dspfparser/linesclassifier/DSPFSpecifications.kt:11]: This property is no longer used

**Review by lanarimarco — CHANGES_REQUESTED** [review]

**Review by lanarimarco — APPROVED** [review]
Ok

**Pull request merged**